### PR TITLE
Updates for operator based configurations.

### PR DESCRIPTION
**Pull request by derekcollison**
Added update to parse and load operator JWTs.
Changed to add in signing keys from operator JWT to list of trusted keys.
Added URL account resolver.
Added account claim updates by system messages.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/accounts.go [hunk 1]
```diff
@@ -925,17 +939,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
...
 
 // Mostly for testing.
 type MemAccResolver struct {
-	sync.Map
+	sm sync.Map
 }
 
-func (m *MemAccResolver) Fetch(pub string) (string, error) {
-	if j, ok := m.Load(pub); ok {
+func (m *MemAccResolver) Fetch(name string) (string, error) {
```
Comment (sasbury) [inline server/accounts.go]: Does this need locking? or does sync map handle?

File: server/accounts.go [hunk 2]
```diff
@@ -925,17 +939,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
...
+		return nil, err
+	}
+	return ur, nil
+}
+
+// Fetch will fetch the account jwt claims from the base url, appending the
+// account name onto the end.
+func (ur *URLAccResolver) Fetch(name string) (string, error) {
+	url := ur.url + name
+	resp, err := ur.c.Get(url)
```
Comment (sasbury) [inline server/accounts.go]: Should we have a timeout and maybe block redirects?

**Comment by derekcollison** [comment]
sync.Map handles


________________________________

**Comment by derekcollison** [comment]
We have a timeout. Not sure about redirects.


________________________________

File: server/accounts.go (hunk 1 above)
Comment (derekcollison) [reply server/accounts.go]: sync.Map handles.

File: server/accounts.go (hunk 2 above)
Comment (derekcollison) [reply server/accounts.go]: We have timeouts already. They are fixed atm but we do set them. Not sure about redirects though. Will take a look.

Comment (derekcollison) [reply server/accounts.go]: I think timeout being set covers us for wonky redirects.

File: server/opts.go [hunk 3]
```diff
@@ -489,12 +492,67 @@ func (o *Options) ProcessConfigFile(configFile string) error {
...
 			}
 			o.LameDuckDuration = dur
-		case "trusted":
+		case "operator", "operators", "roots", "root", "root_operators", "root_operator":
+			opFiles := []string{}
+			switch v.(type) {
+			case string:
+				opFiles = append(opFiles, v.(string))
+			case []string:
+				opFiles = append(opFiles, v.([]string)...)
```
Comment (kozlovic) [inline server/opts.go]: Should you have a `default:` here and report error (like you do for resolver)?

File: server/opts.go [hunk 4]
```diff
@@ -489,12 +492,67 @@ func (o *Options) ProcessConfigFile(configFile string) error {
...
+			}
+			if memResolverRe.MatchString(str) {
+				o.AccountResolver = &MemAccResolver{}
+			} else {
+				items := resolverRe.FindStringSubmatch(str)
+				if len(items) == 2 {
+					url := items[1]
+					if ur, err := NewURLAccResolver(url); err != nil {
+						err := &configErr{tk, fmt.Sprintf("URL account resolver error: %v", err)}
+						errors = append(errors, err)
```
Comment (kozlovic) [inline server/opts.go]: is it ok that there is no `continue` here? we'll fall through and likely fail if o.AccountResolver is nil..

File: server/accounts.go [hunk 5]
```diff
@@ -111,30 +112,40 @@ type exportMap struct {
...
+}
+
+func (a *Account) maxTotalConnectionsReached() bool {
 	if a.mconns != 0 {
 		return len(a.clients)+a.nrclients >= a.mconns
 	}
 	return false
 }
 
+func (a *Account) MaxActiveConnections() int {
```
Comment (kozlovic) [inline server/accounts.go]: Add comment since exported..

File: server/accounts.go [hunk 6]
```diff
@@ -925,17 +951,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
 
 // AccountResolver interface. This is to fetch Account JWTs by public nkeys
 type AccountResolver interface {
-	Fetch(pub string) (string, error)
+	Fetch(name string) (string, error)
```
Comment (kozlovic) [inline server/accounts.go]: Since exported, I would put a comment for each function.

File: server/accounts.go [hunk 7]
```diff
@@ -925,17 +951,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
...
 
 // Mostly for testing.
 type MemAccResolver struct {
-	sync.Map
+	sm sync.Map
 }
 
-func (m *MemAccResolver) Fetch(pub string) (string, error) {
-	if j, ok := m.Load(pub); ok {
+func (m *MemAccResolver) Fetch(name string) (string, error) {
```
Comment (kozlovic) [inline server/accounts.go]: Comment here would be simply: `Fetch implements the AccountResolver interface`. (people can look at the comment in interface def to see what it does)

File: server/accounts.go [hunk 8]
```diff
@@ -925,17 +951,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
...
+	return _EMPTY_, ErrMissingAccount
+}
+
+func (m *MemAccResolver) Store(name, jwt string) error {
+	m.sm.Store(name, jwt)
+	return nil
+}
+
+//
+type URLAccResolver struct {
```
Comment (kozlovic) [inline server/accounts.go]: Exported, so comment?

File: server/accounts.go [hunk 9]
```diff
@@ -925,17 +951,67 @@ func buildInternalNkeyUser(uc *jwt.UserClaims, acc *Account) *NkeyUser {
...
-	if j, ok := m.Load(pub); ok {
+func (m *MemAccResolver) Fetch(name string) (string, error) {
+	if j, ok := m.sm.Load(name); ok {
 		return j.(string), nil
 	}
-	return "", ErrMissingAccount
+	return _EMPTY_, ErrMissingAccount
+}
+
+func (m *MemAccResolver) Store(name, jwt string) error {
```
Comment (kozlovic) [inline server/accounts.go]: comment: `Store implements the AccountResolver interface`

File: server/opts.go (hunk 3 above)
Comment (derekcollison) [reply server/opts.go]: yes will fix.

File: server/opts.go (hunk 4 above)
Comment (derekcollison) [reply server/opts.go]: Would have had double errors so added a continue.

File: server/accounts.go (hunk 5 above)
Comment (derekcollison) [reply server/accounts.go]: +1

File: server/accounts.go (hunk 6 above)
Comment (derekcollison) [reply server/accounts.go]: +1

File: server/accounts.go (hunk 7 above)
Comment (derekcollison) [reply server/accounts.go]: +1

File: server/accounts.go (hunk 8 above)
Comment (derekcollison) [reply server/accounts.go]: Yep, oops.

File: server/accounts.go (hunk 9 above)
Comment (derekcollison) [reply server/accounts.go]: Add one.

**Pull request merged**